x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/SemanticMultiMatchQueryRewriteInterceptor.java

```diff
@@ -0,0 +1,487 @@
+                } else if (parentQuery instanceof BoolQueryBuilder) {
+                    ((BoolQueryBuilder) parentQuery).should(indexSpecificQuery);
+                }
+            }
+        }
+    }
+
+    /**
+     * Adds non-inference queries for non-inference fields per index to the provided query builder.
+     */
+    private void addNonInferenceQueriesPerIndex(
+        QueryBuilder parentQuery,
+        MultiFieldInferenceInfo inferenceInfo,
+        MultiMatchQueryBuilder originalQuery,
+        Map<String, Float> fieldsBoosts
+    ) {
+        Map<String, Set<String>> nonInferenceFieldsPerIndex = inferenceInfo.nonInferenceFieldsPerIndex();
+        for (Map.Entry<String, Set<String>> entry : nonInferenceFieldsPerIndex.entrySet()) {
+            String indexName = entry.getKey();
+            Set<String> nonInferenceFields = entry.getValue();
+
+            MultiMatchQueryBuilder indexSpecificQuery = createNonInferenceQueryForIndex(originalQuery, nonInferenceFields, fieldsBoosts);
+
+            BoolQueryBuilder indexFilteredQuery = new BoolQueryBuilder();
+            indexFilteredQuery.must(indexSpecificQuery);
+            indexFilteredQuery.filter(new TermsQueryBuilder(IndexFieldMapper.NAME, List.of(indexName)));
+
+            if (parentQuery instanceof DisMaxQueryBuilder) {
+                ((DisMaxQueryBuilder) parentQuery).add(indexFilteredQuery);
+            } else if (parentQuery instanceof BoolQueryBuilder) {
+                ((BoolQueryBuilder) parentQuery).should(indexFilteredQuery);
+            }
+        }
+    }
+
+    private QueryBuilder buildBestFieldsSemanticQuery(
+        MultiMatchQueryBuilder originalQuery,
+        Map<String, Float> fieldsBoosts,
+        Set<String> inferenceFields,
+        String queryValue
+    ) {
+        DisMaxQueryBuilder disMaxQuery = QueryBuilders.disMaxQuery();
+        for (String fieldName : inferenceFields) {
+            disMaxQuery.add(createSemanticQuery(fieldName, queryValue, fieldsBoosts, false));
+        }
+        // Apply tie_breaker if specified
+        if (originalQuery.tieBreaker() != null) {
+            disMaxQuery.tieBreaker(originalQuery.tieBreaker());
+        }
+        disMaxQuery.boost(originalQuery.boost());
+        disMaxQuery.queryName(originalQuery.queryName());
+        return disMaxQuery;
+    }
+
+    /**
+     * Builds a most_fields query for pure semantic fields using BoolQueryBuilder.
+     */
+    private QueryBuilder buildMostFieldsSemanticQuery(
+        MultiMatchQueryBuilder originalQuery,
+        Map<String, Float> fieldsBoosts,
+        Set<String> inferenceFields,
+        String queryValue
+    ) {
+        BoolQueryBuilder boolQuery = new BoolQueryBuilder();
+        for (String fieldName : inferenceFields) {
+            boolQuery.should(createSemanticQuery(fieldName, queryValue, fieldsBoosts, false));
+        }
+        boolQuery.minimumShouldMatch("1");
```

Comment:
We don't want to hard code this, because it can be specified in the multi match query.